### PR TITLE
EKF3: allow earth-frame fields to be estimated with an origin but no GPS

### PR DESCRIPTION
**Pull request by Williangalvani**
This is my attempt at fixing https://github.com/ArduPilot/ardupilot/issues/18229

Here's a log using it:
[00000228.zip](https://github.com/ArduPilot/ardupilot/files/13502545/00000228.zip)
I am using a script to set the origin so we can use the earth field database.

I can see that XKF2 fields are all populated for magnetic field in body frame and earth frame.

image

I am not sure of what else to look at, though. magnetic field innovations seem bad, but at least they exist?
image

### CODE REVIEW

**Comment by rmackay9** [comment]
Hi, if the issue is the earth magnetic field (aka WMM) is not being estimated then the EKF also shouldn't be trying to use the WWM to reject the compass, instead it should accept it I think.

As a side note, it's naughty that the EKF is using the GPS status at all.  This presumably means the same problem could happen when using a non-GPS navigation source while also using a compass which is very possible (e.g. Beacons).

**Comment by lixiaowei123** [comment]
af113c5725f37d8b81e856b1ebd5435
I have been saying that there is a bug in the estimation of EKF3 acceleration velocity increment offset

**Comment by lixiaowei123** [comment]
@rmackay9 You can view all the latest version logs to see if this is sometimes very large, and normally it should not exceed 0.3

File: libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp [hunk 1]
```diff
@@ -548,6 +548,23 @@ void NavEKF3_core::readGpsData()
...
+        const auto &compass = dal.compass();
+        if (compass.have_scale_factor(magSelectIndex) &&
+            compass.auto_declination_enabled()) {
+            getEarthFieldTable(EKF_origin);
+            if (frontend->_mag_ef_limit > 0) {
+                // initialise earth field from tables
+                stateStruct.earth_magfield = table_earth_field_ga;
+            }
+        }
+        return;
```
Comment (tridge) [inline libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp:565]: why the early return? this seems to defeat some of the code
maybe only return if the xy position source type is none?

**Review by tridge — APPROVED** [review]

File: libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp [hunk 2]
```diff
@@ -548,6 +548,23 @@ void NavEKF3_core::readGpsData()
     // check for new GPS data
     const auto &gps = dal.gps();
 
+    Vector3f magNED;
+    getMagNED(magNED);
+    if (gps.status() == AP_DAL_GPS::NO_GPS && validOrigin && !stateStruct.quat.is_zero()) {
```
Comment (tridge) [inline libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp:553]: maybe this should be in the function that sets the origin?

Comment (tridge) [reply libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp:553]: NavEKF3_core::setOrigin()

**Comment by rmackay9** [comment]
As discussed on the dev call we should investigate why the EKF3 is unable to fuse the compass when it doesn't have a position estimate.  It really should be able to although it won't be as accurate.

File: libraries/AP_NavEKF3/AP_NavEKF3_Control.cpp [hunk 3]
```diff
@@ -682,6 +682,24 @@ bool NavEKF3_core::setOrigin(const Location &loc)
     // define Earth rotation vector in the NED navigation frame at the origin
     calcEarthRateNED(earthRateNED, EKF_origin.lat);
     validOrigin = true;
+
+    // this section is similar to what is in AP_NavEKF3_Measurements.cpp's NavEKF3_core::readGpsData()
+    // but we do want to populate the WMM table even if we don't have a GPS at all
+    Vector3f magNED;
```
Comment (tridge) [inline libraries/AP_NavEKF3/AP_NavEKF3_Control.cpp:688]: 2 unused lines?

File: libraries/AP_NavEKF3/AP_NavEKF3_Control.cpp [hunk 4]
```diff
@@ -682,6 +682,24 @@ bool NavEKF3_core::setOrigin(const Location &loc)
...
+    getMagNED(magNED);
+    if (!stateStruct.quat.is_zero()) {
+        alignMagStateDeclination();
+        const auto &compass = dal.compass();
+        if (compass.have_scale_factor(magSelectIndex) &&
+            compass.auto_declination_enabled()) {
+            getEarthFieldTable(EKF_origin);
+            if (frontend->_mag_ef_limit > 0) {
+                // initialise earth field from tables
+                stateStruct.earth_magfield = table_earth_field_ga;
```
Comment (tridge) [inline libraries/AP_NavEKF3/AP_NavEKF3_Control.cpp:698]: I think you need to fill in these: ```
    // earth field from WMM tables
    bool have_table_earth_field;   // true when we have initialised table_earth_field_ga
    Vector3F table_earth_field_ga; // earth field from WMM tables
    ftype table_declination;       // declination in radians from the tables
```

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_NavEKF3/AP_NavEKF3_Control.cpp [hunk 5]
```diff
@@ -682,6 +682,22 @@ bool NavEKF3_core::setOrigin(const Location &loc)
...
     calcEarthRateNED(earthRateNED, EKF_origin.lat);
     validOrigin = true;
+
+    // this section is similar to what is in AP_NavEKF3_Measurements.cpp's NavEKF3_core::readGpsData()
+    // but we do want to populate the WMM table even if we don't have a GPS at all
+    if (!stateStruct.quat.is_zero()) {
+        alignMagStateDeclination();
+        const auto &compass = dal.compass();
+        if (compass.have_scale_factor(magSelectIndex) &&
+            compass.auto_declination_enabled()) {
```
Comment (tridge) [inline libraries/AP_NavEKF3/AP_NavEKF3_Control.cpp:692]: don't need to check auto_declination_enabled

File: libraries/AP_NavEKF3/AP_NavEKF3_Control.cpp [hunk 6]
```diff
@@ -682,6 +682,22 @@ bool NavEKF3_core::setOrigin(const Location &loc)
     // define Earth rotation vector in the NED navigation frame at the origin
     calcEarthRateNED(earthRateNED, EKF_origin.lat);
     validOrigin = true;
+
+    // this section is similar to what is in AP_NavEKF3_Measurements.cpp's NavEKF3_core::readGpsData()
+    // but we do want to populate the WMM table even if we don't have a GPS at all
+    if (!stateStruct.quat.is_zero()) {
+        alignMagStateDeclination();
+        const auto &compass = dal.compass();
+        if (compass.have_scale_factor(magSelectIndex) &&
```
Comment (tridge) [inline libraries/AP_NavEKF3/AP_NavEKF3_Control.cpp:691]: could be a common function with the code in readGpsData()

Comment (tridge) [reply libraries/AP_NavEKF3/AP_NavEKF3_Control.cpp:691]: maybe setEarthFieldFromLocation()

**Comment by rmackay9** [comment]
We discussed on the July 2nd dev call and decided that we should move the shared code to a new function and then call it from within setOrigin and readGPSData.

**Review by rmackay9 — APPROVED** [review]

**Review by tridge — APPROVED** [review]

**Pull request merged**